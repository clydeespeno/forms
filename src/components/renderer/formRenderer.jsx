import React from 'react'
import map from 'ramda/src/map'

const renderField = (renderComponent) => (field) => ({
  field,
  component: renderComponent(field)
})

const renderLayoutOf = (mapComponent, renderComponent) => {
  const renderFields = map(renderField(renderComponent))

  // eslint-disable-next-line react/prop-types
  return ({fields, ...layout}) => {
    const Layout = mapComponent(layout)

    return <Layout layout={layout} fields={renderFields(fields)}/>
  }
}

const renderComponentOf = (mapComponent) => {
  return ({props, ...component}) => {
    const Component = mapComponent(component)

    return <Component {...props}/>
  }
}

const renderForm = (mapComponent) => {
  const render = (component) => {
    if (component.type === 'layout') {
      return renderLayout(component)
    }

    return renderComponent(component)
  }

  const renderComponent = renderComponentOf(mapComponent)
  const renderLayout = renderLayoutOf(mapComponent, render)

  return render
}

export const formRenderer = (mapComponent) => {
  const render = renderForm(mapComponent)
  return (form) => {
    return render(form)
  }
}
