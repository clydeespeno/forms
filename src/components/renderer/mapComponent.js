export const mapComponentOf = (mapperOf) => ({type, componentId}) =>
  mapperOf(type)(componentId)
