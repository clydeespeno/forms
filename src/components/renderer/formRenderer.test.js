import TestRenderer from 'react-test-renderer'
import {formRenderer} from './formRenderer'
import {form, formWithFields} from './fixture/form.fixture'
import {mapComponent} from './fixture/renderer.test'

describe('formRenderer', () => {
  it('renders an empty layout', () => {
    expect(testRender(form()).toJSON()).toMatchSnapshot()
  })

  it('renders a layout with fields', () => {
    expect(testRender(formWithFields()).toJSON()).toMatchSnapshot()
  })

  const testRender = (form) => TestRenderer.create(renderForm(form))

  const renderForm = formRenderer(mapComponent)
})
