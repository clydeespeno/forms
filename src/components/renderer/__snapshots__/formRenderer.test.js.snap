// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formRenderer renders a layout with fields 1`] = `
<div
  className="layout"
>
  <div
    id="input-input-input"
  >
    <div>
      input
    </div>
    <div
      className="component"
    >
      <div>
        inputed
      </div>
    </div>
  </div>
  <div
    id="select-select-select"
  >
    <div>
      select
    </div>
    <div
      className="component"
    >
      <div>
        selected
      </div>
    </div>
  </div>
</div>
`;

exports[`formRenderer renders an empty layout 1`] = `
<div
  className="layout"
/>
`;
